### DISCUSSION TITLE
Element callback update

### DISCUSSION THREAD

**Discussion by morningstarsabrina** [zauberzeug/nicegui#345] [Ideas / Feature Requests]
Hello,

Is it possible to add set_callback method to elements such as button so we can change or add the callback later ?

thanks

**Comment by morningstarsabrina** [comment]
can somebody help me to achieve this ?

**Comment by rodja** [comment]
Why would you like to do this? You can register a callback function to the button and then inside the function decide what to do. Why change the callback itself? Otherwise each element has a list `._event_listeners` which you could manipulate...

**Reply by morningstarsabrina** [reply]
I want to do this because I want to use separated files design. defining element in one file and add callback in other files when I have logic functions.
could you show me an example about   `._event_listeners`

thanks

**Reply by rodja** [reply]
I see. You don't want to change the callback. You just want to add a callback. That can be achieved with the generic `.on` event registration. Here is a super small example:

```py
b = ui.button('Click')
ui.button('Add callback', on_click=lambda: b.on('click', lambda: ui.notify('Clicked!')))
```

**Reply by morningstarsabrina** [reply]
yes I want to add call back.
I hope in future there will be a cleaner method to add callbacks, but currently this is more than enough.
I come from Qt framework when you can simply use methods  such as 'connect' to add call back and 'disconnect' to stop listening. you can even add more than 1 callback.
thank you very much for your help !

**Reply by rodja** [reply]
With `.on` you can already add multiple callbacks to one element. But yes, removing/updating is not so easy right now. We will may implement something in the future. Thanks for sharing your requirement with us.

**Reply by morningstarsabrina** [reply]
thank you for accepting suggestions in a positive way !

**Comment by morningstarsabrina** [comment]
@rodja 

so I'm trying the example you gave me
```py
b = ui.button('Click')
ui.button('Add callback', on_click=lambda: b.on('click', lambda: ui.notify('Clicked!')))
```
I noticed now it create another button with text 'Add callback' that I have to click on it first to make b button work.
I'm doing something wrong ?
What I'm trying to do is to add a call back the 'b' button without creating another button

**Reply by rodja** [reply]
The second button was only meant for demonstration. Use it like this:

In your design file:
```py
b = ui.button('Click')
```

In your business logic:
```py
def some_on_click_handler():
    print('clicked')

b.on('click', some_on_click_handler)
```

**Reply by morningstarsabrina** [reply]
ok thanks.
by the way for me print() work only if I put ` print('clicked', flush = True)` is this normal ? is there a better way without putting flush ?

**Reply by rodja** [reply]
Yes, flushing is required when using `print`. Most of the time we work with the [Python logging module](https://docs.python.org/3/library/logging.html).

**Reply by morningstarsabrina** [reply]
thanks for the tip
